src/peft/tuners/lora/layer.py

```diff
@@ -250,6 +253,56 @@ def pissa_init(self, adapter_name, init_lora_weights):
         weight = transpose(weight.to(dtype), self.fan_in_fan_out)
         self.get_base_layer().weight.data = weight
 
+    def corda_init(self, adapter_name, init_lora_weights):
+        linear = self.get_base_layer()
+        weight = linear.weight
+        dtype = weight.dtype
+        if dtype not in [torch.float32, torch.float16, torch.bfloat16]:
+            raise TypeError(
+                "Please initialize CorDA under float32, float16, or bfloat16. "
+                "Subsequently, re-quantize the residual model to help minimize quantization errors."
+            )
+        weight = weight.to(torch.float32)
+        out_dim = weight.data.size(0)
+        in_dim = weight.data.size(1)
+
+        # Calculate WC from covariance matrix
+        if not hasattr(linear, "eigens"):
+            raise ValueError("eigens attribute not found, it's expected to be set by the pre-injection hook.")
+        eigens = linear.eigens
+        U = eigens.U_WC
+        S = eigens.S_WC
+        V = eigens.V_WC
+        r = self.r[adapter_name]
+
+        # nan or inf check
+        if torch.isnan(S).any() or torch.isinf(S).any():
+            raise ValueError("nan or inf in S")
+        if torch.isnan(U).any() or torch.isinf(U).any():
+            raise ValueError("nan or inf in U")
+        if torch.isnan(V).any() or torch.isinf(V).any():
+            raise ValueError("nan or inf in V")
+
+        # Sanity check
+        if U.size(0) != out_dim or U.size(1) != r:
+            raise ValueError(f"U size mismatch: {U.size()} vs. ({out_dim}, {r})")
+        if S.size(0) != r:
+            raise ValueError(f"S size mismatch: {S.size()} vs. ({r},)")
+        if V.size(0) != in_dim or V.size(1) != r:
+            raise ValueError(f"V size mismatch: {V.size()} vs. ({in_dim}, {r})")
```

Comment:
Similar comment as above: What should the user do if they encounter this error? Is it something they can mitigate by setting other hyper-parameters for CorDA or using a different dataset sample?